### DISCUSSION TITLE
Multi-tenancy using partition key in a collection

### DISCUSSION THREAD

**Discussion by gtvracer** [milvus-io/milvus#37746] [Q&A and General discussion]
Hello, 

Reading the Milvus docs about multi-tenancy, it seems using partition key in one collection per database was the most flexible for me.

But in one part of the doc, it states:  If you have set a field as the partition key in a collection, Milvus creates at least 64 partitions along with the collection.
The sample code in Milvus docs:

schema = MilvusClient.create_schema(
    auto_id=False,
    enable_dynamic_field=True,
    partition_key_field="color",
    num_partitions=64 # Number of partitions. Defaults to 64.
)

And in another section: 

Milvus 2.2.9 introduces a new feature named partition key. Upon the creation of a collection, nominate a tenant field and make it the partition key field. Milvus will store entities in a partition according to the values in the partition key field. 

Maximum number of tenants is over 10M!

So my question is, just leave the num_partitions value at 64 and forget about it and let Milvus take care of the partitioning the collection?  Also, how do we specify this partition key in a MilvusClient.search() api?   All I could find was their doc indicating to use it as a filter.   But then, where is this advantage of the partition key if it's a simple filtering?

res = client.search(
    collection_name="test_collection",
    data=query_vectors,
    filter="color == 'green'",
    search_params={"metric_type": "L2", "params": {"nprobe": 10}},
    output_fields=["id", "color_tag"],
    limit=3
)

**Comment by xiaofan-luan** [comment]
Hi @gtvracer 
when you do search, you need to speicify the partitionkey.
For example, if parititonkey is userID, then specify userID=1 will limit you request to only one single tenant

**Reply by gtvracer** [reply]
In the filter, just like any other filter argument?  There is no differentiation between a simple filter and a partition key, which seems more specialized.

**Reply by xiaofan-luan** [reply]
> In the filter, just like any other filter argument? There is no differentiation between a simple filter and a partition key, which seems more specialized.

yes, milvus compiler will figure out which is the partition key and use different route stratgy

**Comment by yhmo** [comment]
**How partition-key works?**

1. Assume a collection has a partition key named "name", and you create the collection with num_partitions=5, milvus automatically create 5 partitions for you, their name is partition_0, partition_1, partition_2, partition_3, partition_4
2. Now you insert an entity into the collection:
`collection.insert({"id": 1, "name": "abc", "vector": [xxx]})`
3. The value "abc" is hashed to a value, let's say 123456. Then 123456%5=1, the entity is stored into the partition_1
4. You insert another entity into the collection:
`collection.insert({"id": 2, "name": "xyz", "vector": [xxx]})`
5. The value "xyz" is hashed to a value, let's say 45678. Then 45678%5=3, the entity is stored into the partiton_3
6. You insert more and more entities, no matter how many entities you insert, they are always inserted into partition_0 to partition_4
7. You search with filter expression 
```
res = client.search(
filter="name == 'xyz'",
)
```
8. Milvus hashes the value "xyz" to 45678. Since 45678%5=3, milvus knows the item must in partition_3. So, it only searches partition_3, other partitions are skipped to improve search performance significantly.


**Can I delete or add a partition with partition key exists?**
No, you can't.
Because the Hash+Mod algorithm doesn't allow changing the number of partitions.
1. Assume you added a partition_5, now there are 6 partitions in the collection.
2. You insert a new entity with name="abc", "abc" is hashed to 123456. Then 123456%6=0, the entity is stored into partition_0.
3. Now both partition_1 and partition_0 have entities with name="abc"
4. You search with a filter expression filter="name=='abc'", "abc" is hashed to 123456 and 123456%6=0. It only searches the partition_0, the entities with name="abc" in partition_1 is missed.


**Does the partition key support any filter expression?**
No, it only supports expressions like "name='xxx'" or "name in ['xxx', 'xxx', 'xxx']"
Assume you search with this expression:
filter = "name < 'abc'"
The hash value of "abc" cannot be used to compare greater or smaller. So, the search engine will fall back to search the entire collection.
